index.ts

```diff
@@ -64,76 +64,126 @@ export interface Service {
-        methodTimeout = 60000,
-        help,
-        paramNames = [],
-      } = method;
-
-      return {
-        methodName,
-        paramNames,
-        methodTimeout,
-        help: help || methodName + " method",
-      };
-    }),
-  };
 
-  return (
-    req: { path: string; method: string; body: unknown },
-    res: { json: (body: unknown) => void },
-    next: (err?: Error) => void
-  ) => {
-    const methodName: string = req.path.slice(1); // remove leading slash
+      if (req.method !== "POST" || !methods.includes(methodName)) {
+        return next();
+      }
 
-    if (req.method === "GET" && req.path === "/") {
-      return res.json(meta);
-    }
+      const requestMeta = { handler: `${serviceName}.${methodName}` };
+      const end = requestDuration.startTimer(requestMeta);
 
-    if (req.method === "GET" && methods.includes(methodName)) {
-      return res.json(meta.interfaces.find((i) => i.methodName === methodName));
-    }
+      requestCount.inc(requestMeta);
 
-    if (req.method !== "POST" || !methods.includes(methodName)) {
-      return next();
-    }
+      return Promise.resolve()
+        .then(() => this.service[methodName].call(this.service, req.body))
+        .then((result) => (console.log("result", result), res.json(result)))
+        .catch((err) => {
+          failureCount.inc(Object.assign({ type: err.type }, requestMeta));
+          next(new RpcError(serviceName, methodName, err));
+        })
+        .finally(end);
+    };
+  }
 
-    const requestMeta = { handler: `${serviceName}.${methodName}` };
-    const end = requestDuration.startTimer(requestMeta);
+  toWellKnown() {
+    return {
+      service: this.serviceDetails.service,
+      help: this.serviceDetails.help,
+      path: `/rpc/${this.serviceDetails.service}`,
+    };
+  }
+}
 
-    requestCount.inc(requestMeta);
+export class Registry {
+  // TODO: Fix the RegisteredService type from 'any'
+  private registeredServices: Record<string, RegisteredService<any>> = {};
 
-    return Promise.resolve()
-      .then(() => service[methodName].call(service, req.body))
-      .then((result) => (console.log("result", result), res.json(result)))
-      .catch((err) => {
-        failureCount.inc(Object.assign({ type: err.type }, requestMeta));
-        next(new RpcError(serviceName, methodName, err));
-      })
-      .finally(end);
-  };
+  register<S extends Service>(
+    service: S,
+    serviceDetails: ServiceDetails<S>
+  ): RegisteredService<S> {
+    const rs = new RegisteredService<S>(service, serviceDetails);
+
+    this.registeredServices[serviceDetails.service] = rs;
+
+    return rs;
+  }
+
+  createWellKnownHandler(): RequestHandler {
+    return (
+      req: { path: string; method: string; body: unknown },
+      res: { json: (body: unknown) => void }
+    ) => {
+      const requestMeta = { handler: `well-known-services` };
+      const end = requestDuration.startTimer(requestMeta);
+
+      requestCount.inc(requestMeta);
```

Comment:
Was undecided before, but now think that the metrics aren't very useful here.

I reckon get rid of them.

Reply:
done